src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp

```diff
@@ -360,31 +360,39 @@ bool ShenandoahHeap::is_in(const void* p) const {
 inline bool ShenandoahHeap::is_in_active_generation(oop obj) const {
   if (!mode()->is_generational()) {
     // everything is the same single generation
+    assert(is_in(obj), "Otherwise shouldn't return true below");
     return true;
   }
 
-  if (active_generation() == nullptr) {
-    // no collection is happening, only expect this to be called
+  ShenandoahGeneration* const gen = active_generation();
+
+  if (gen == nullptr) {
+    // no collection is happening: only expect this to be called
     // when concurrent processing is active, but that could change
     return false;
   }
 
   assert(is_in(obj), "only check if is in active generation for objects (" PTR_FORMAT ") in heap", p2i(obj));
-  assert((active_generation() == (ShenandoahGeneration*) old_generation()) ||
-         (active_generation() == (ShenandoahGeneration*) young_generation()) ||
-         (active_generation() == global_generation()), "Active generation must be old, young, or global");
+  assert(gen == (ShenandoahGeneration*)old_generation() ||
```

Comment:
Done... Seems more patently tautological now in the new form, but I'll let it stay in this form for now. Let me know if it can or should be removed entirely.